### PR TITLE
Rmvanarse/gazebosim/scoring plugins dev

### PR DESCRIPTION
**Pull request by rmvanarse**
## Scoring plugin implemented with ignition

**Files added**
- `ScoringPlugin.hh`
- `ScoringPlugin.cc`
- `DummyPlugin.hh`
- `DummyPugin.cc`

The dummy plugins contain skeleton code that can be used for implementing other plugins. The scoring plugins have been structured similarly to the corresponding files in VRX. The main modifications are:
- Functions such as Load() and Update() have been replaced by Configure() and ISystemPreUpdate/ ISystemPostUpdate
- World/ model objects are now entities. Their attributes in Gazebo classic implementations are now components, accessed via the entity component system or a World/Model wrapper to the entities
- ROS publishers and subscribers have been replaced with ignition::transport and ROS messages have been replaced with ignition::msgs.
- Several functions for accessing or setting members have changed (e.g. GetLatitudeReference(), GetHeadingOffset() ). Some functions now require extra arguments such as the ECM or simTime (e.g. UpdateTime() ).
- Changes in data types - The main one is the replacement of ignition::common::Time (deprecated) with std::chrono::duration<double>. 

The following still needs to be done:
- Making time variables consistent in type or unit. Currently std::chrono::duration, std::chrono::steady_clock, ignition::msgs::Time and double are all being used. Additionally, ignition::common::Time, builtin_interfaces::msgs::Time and rclcpp::Time have been removed. 
- Completing the implementation of the PostUpdate function (Accessing the vehicle model and name)
- Code cleanup

### CODE REVIEW

File: vrx_ign/CMakeLists.txt [hunk 1]
```diff
@@ -21,9 +22,10 @@ set(IGN_PLUGIN_VER ${ignition-plugin1_VERSION_MAJOR})
 find_package(ignition-rendering6 REQUIRED)
 set(IGN_RENDERING_VER ${ignition-rendering6_VERSION_MAJOR})
 find_package(sdformat12 REQUIRED)
+find_package(ignition-physics4 REQUIRED)
```
Comment (caguero) [inline vrx_ign/CMakeLists.txt:25]: This should be `physics5` as this is the version in Gazebo Fortress. See [this repository](https://github.com/gazebo-tooling/gazebodistro/blob/master/collection-fortress.yaml) for all the library versions in each Gazebo distribution. Change the next line too.

**Comment by caguero** [comment]
Nit: For code consistency with the Gazebo style, could you use the delimiter `//////////////////////////////////////////////////` above each function in the `.cc`. See other files as an example.

File: vrx_ign/src/ScoringPlugin.hh [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        /// \brief Get the number of WAM-V collisions.
+        protected: unsigned int GetNumCollisions() const;
+
+        /// \brief Callback executed at every world update.
+        public: void PostUpdate(
+                    const ignition::gazebo::UpdateInfo &_info,
+                    const ignition::gazebo::EntityComponentManager &_ecm) override;
+
+        /// \brief Update all time-related variables.
+        std::chrono::duration<double> simTime;
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.hh:114]: Remove?

File: vrx_ign/src/ScoringPlugin.hh [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        private: float collisionBuffer = 3.0;
+
+        /// \brief Collisions counter.
+        private: int collisionCounter = 0;
+
+        /// \brief Collision list.
+        private: std::vector<std::string> collisionList;
+
+        /// \brief Collisions timestamps.
+        private: std::vector<std::chrono::duration<double> > collisionTimestamps;
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.hh:238]: Nit: Remove the space between the two `>`.

File: vrx_ign/src/ScoringPlugin.hh [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+        /// \brief Time at which the last message was sent.
+        private: std::chrono::duration<double> lastStatsSent;
+
+        /// \brief The task state.
+        private: std::string taskState = "initial";
+        
+        /// \brief The next task message to be published.
+        ignition::msgs::Param_V taskMsg;
+        ignition::msgs::Any taskMsgName, taskMsgState,
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.hh:251]: Nit: Following the Gazebo convention, don't group member variables. Instead, declare them in separate lines.

File: vrx_ign/src/ScoringPlugin.hh [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        /// \brief ROS Contact Msg.
+        // private: vrx_ros::msg::Contact contactMsg; //Replaced with ignition messages
+        private: ignition::msgs::Contact contactMsg;
+
+        /// \brief The name of the joints to be dettached during ReleaseVehicle().
+        private: std::vector<std::string> lockJointNames;
+
+        /// \brief Publisher for the task state.
+        //Replaced with ignition messages
+        protected: std::unique_ptr<ignition::transport::Node::Publisher> taskPub;
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.hh:266]: Nit: Split the lines that go beyond 80 characters.

File: vrx_ign/src/ScoringPlugin.hh [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+        /// \brief Score in case of timeout - added for Navigation task
+        private: double timeoutScore = -1;
+
+        /// \brief Whether to shut down after last gate is crossed.
+        private: bool perPluginExitOnCompletion = true;
+
+        /// \brief Number of WAM-V collisions.
+        private: unsigned int numCollisions = 0u;
+
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.hh:285]: Nit: remove empty line.

File: vrx_ign/src/ScoringPlugin.cc [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+    taskMsgRunningTime.set_type(ignition::msgs::Any_ValueType::Any_ValueType_TIME);
+    taskMsgElapsedTime.set_type(ignition::msgs::Any_ValueType::Any_ValueType_TIME);
+    taskMsgRemainingTime.set_type(ignition::msgs::Any_ValueType::Any_ValueType_TIME);
+    taskMsgTimedOut.set_type(ignition::msgs::Any_ValueType::Any_ValueType_BOOLEAN);
+    taskMsgNumCollisions.set_type(ignition::msgs::Any_ValueType::Any_ValueType_INT32);
+    taskMsgScore.set_type(ignition::msgs::Any_ValueType::Any_ValueType_DOUBLE);
+
+    taskMsgName.set_string_value(this->taskName);
+
+    ignition::msgs::Time readyTimeMsg, runningTimeMsg;
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.cc:57]: Nit: Split into two separate lines.

File: vrx_ign/src/ScoringPlugin.cc [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+    
+    auto ready_sec = std::chrono::duration_cast<std::chrono::seconds>(this->readyTime);
+    auto ready_nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(this->readyTime - ready_sec);
+    readyTimeMsg.set_sec(ready_sec.count());
+    readyTimeMsg.set_nsec(ready_nsec.count());
+    this->taskMsgReadyTime.set_allocated_time_value(&readyTimeMsg);
+
+    runningTimeMsg.set_sec(this->runningTime.count());
+
+    auto running_sec = std::chrono::duration_cast<std::chrono::seconds>(this->runningTime);
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.cc:67]: I recommend to create a helper function to set the time, as it appears many times.

File: vrx_ign/src/ScoringPlugin.cc [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+    // std::string worldName = _ecm.Component<ignition::gazebo::components::Name>(this->world);
+    std::string worldName = this->world.Name(_ecm).value();
+    
+    std::string collisionTopic =
+        std::string("/gazebosim/") + worldName + std::string("/physics/contacts");
+    if (!gzNode->Subscribe(collisionTopic, &ScoringPlugin::OnCollisionMsg, this))
+    {
+        std::cerr << "Error subscribing to [" << collisionTopic << "]" << std::endl;
+    }
+
```
Comment (caguero) [inline vrx_ign/src/ScoringPlugin.cc:114]: The code in VRX verifies the `VRX_DEBUG` environment variable to set `this->debug`. Should we add it?

**Comment by caguero** [comment]
Could you provide instructions for testing your PR please?

**Review by caguero — CHANGES_REQUESTED** [review]

File: vrx_ign/CMakeLists.txt [hunk 10]
```diff
@@ -4,6 +4,7 @@ project(vrx_ign)
 
 find_package(ament_cmake REQUIRED)
 find_package(ament_cmake_python REQUIRED)
+find_package(rclcpp REQUIRED)
```
Comment (caguero) [inline vrx_ign/CMakeLists.txt:7]: Do we need this?

**Review by caguero — CHANGES_REQUESTED** [review]

**Comment by caguero** [comment]
@rmvanarse , could you please fix the conflicts?

**Comment by caguero** [comment]
@rmvanarse , in order to unblock this pull request, I suggest:

* Document all the issues, ToDos and questions in a separate issue.
* Polish the part that's working and wrap it up.
* We'll create a separate PR to address the outstanding issues of this plugin.

Does it work for you?

**Comment by rmvanarse** [comment]
Sure! Will do it 👍🏼

**Comment by caguero** [comment]
Replaced with #501 .

**Pull request closed without merging**